Add mediaplatform, mediaplatform_jwp applicationand legacysms models referencing said

rjw57 opened the pull request:
> Note: this was re-opened from #136 after re-naming the ``jwplatform`` application to ``mediaplatform_jwp``.

This PR adds a new ``mediaplatform`` application and associated models (#120), a ``mediaplatform_jwp`` application and associated models (#129) and models referencing the ``mediaplatform`` models to the ``legacysms`` application (#121). It also expands the ACL scheme from (#127) into a new "permission" model but does not actually provide any of the logic specified in #127.

No indexes have been added to the models yet. We'll add them as we need them.

See the individual commit messages for more detail.

Closes #120. Closes #121. Closes #129.

abrahammartin commented on mediaplatform/models.py:
```diff
@@ -0,0 +1,146 @@
+
+    """
+    #: Primary key
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
This should be shorter to reference it in the url or are you thinking in doing it at runtime?

rjw57 replied:
I think the URL encoding is a separate problem. Base64 encoding the UUIDs should lead to URLs which look like https://alpha.media.cam.ac.uk/media/IeGheechaegool9ahth7ee which I think is probably OK. We could use a random 64-bit integer instead which would lead to URLs like https://alpha.media.cam.ac.uk/media/ool2ieFe0e. I'm neutral on which we choose so I'm happy to do whatever you'd prefer.

rjw57 replied:
As a separate issue, I think it'd be worth running the Base64 encoded ids through, e.g., https://github.com/Volnus/YouTube-Blacklist to avoid accidentally having media with URLs containing bad words.

rjw57 replied:
> I'm neutral on which we choose so I'm happy to do whatever you'd prefer.

Actually, I'm coming down on the side of a random 64-bit integer as providing a nicer compromise so I think I'll re-work this to use random 64-bit integer ids. I'll work the Birthday Paradox probabilities first though.

Alternatively, we could use an approach like https://wiki.postgresql.org/wiki/Pseudo_encrypt or https://wiki.postgresql.org/wiki/Pseudo_encrypt_constrained_to_an_arbitrary_range but I think that's becoming needlessly complex.

rjw57 replied:
So, according to the table at https://en.wikipedia.org/wiki/Birthday_problem#Probability_table, to get a 1 in a million chance of collision with 64-bit ids we'd need around 6 million media items/collections. For 128-bit UUIDs, that becomes 2.6 x 10^16.

Further, collision only actually matters if we use 64-bit ids for more than one table (e.g. media items and collections). Within a table, Postgres saves us from collision. With that the probability that a) a collision occurs and b) it happens to be between a collection and a media item probably becomes tolerably small.

abrahammartin commented on mediaplatform/models.py:
```diff
@@ -0,0 +1,146 @@
+
+    """
+    #: Primary key
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```
same comment as with media

rjw57 replied:
See reply above.

abrahammartin commented on mediaplatform/models.py:
```diff
@@ -0,0 +1,146 @@
+    #: handle the (rare) case that a UUID in the list does not correspond to a current video.
+    #: YouTube, as an example, has this problem as well since videos in playlists are sometimes
+    #: replaced by a "deleted video" placeholder.
+    media_items = pgfields.ArrayField(
```
why not just a foreingkey in media_item? Is it because we may want to preserve the order? I think is worth adding it to the comments.

rjw57 replied:
1) If a foreign key in media_item then a media item can be in one and only one collection which seems a needless restriction.
2) If a foreign key in media_item then there is no ordering

Looking at the way the SMS currently does it, there is a separate table mapping media to collections with an index field. Although this would work, it seems ugly. If you're unhappy with losing the foreign key constraint, I'd use a many-to-many mapping table like this instead.

abrahammartin commented on legacysms/models.py:
```diff
@@ -1,3 +1,29 @@
+
+    #: Corresponding :py:class:`mediaplatform.MediaItem`. Accessible from the
+    #: :py:class:`mediaplatform.MediaItem` model via the ``sms`` field.
+    item = models.OneToOneField('mediaplatform.MediaItem', related_name='sms',
```
currently we can't delete videos from the SMS via this method, models.CASCADE will make it disappear form the website but not from the SMS. It's a valid option but we need to think if that then will break our sync and it will try it to upload it again

rjw57 replied:
At no point will the UMP become the source of truth for the SMS so at no point will deleting a video from UMP be expected to delete it from the SMS. Rather, when the SMS ceases to be the source of truth, a video deleted on the UMP will remain on the SMS because they are unconnected systems. While the SMS is still the source of truth for the UMP, "deleting a video" should be done on the SMS and then synchronised to the UMP.

As I understand it, therefore, this is working as designed.

abrahammartin commented on legacysms/models.py:
```diff
@@ -1,3 +1,29 @@
+    id = models.BigIntegerField(primary_key=True, editable=False,
+                                help_text='Legacy SMS collection id')
+
+    #: Corresponding :py:class:`mediaplatform.Collection`. Accessible from the
```
same comment as with media

rjw57 replied:
Same reply :).

abrahammartin commented on mediaplatform_jwp/models.py:
```diff
@@ -0,0 +1,18 @@
+
+    #: Corresponding :py:class:`mediaplatform.MediaItem`. Accessible from the
+    #: :py:class:`mediaplatform.MediaItem` model via the ``jwp`` field.
+    item = models.OneToOneField('mediaplatform.MediaItem', related_name='jwp',
```
same comment as with sms legacy

rjw57 replied:
A similar reply to the one about SMS but with the subtlety that it is unclear to me if it is ever valid to have a video on JWPlatform which has no corresponding ``mediaplatform.MediaItem`` entry. I've assumed not but my assumption could be mistaken.

abrahammartin commented on mediaplatform_jwp/models.py:
```diff
@@ -0,0 +1,18 @@
+                                on_delete=models.CASCADE)
+
+    #: The updated timestamp from JWPlatform. Used to determine which items need updating.
+    updated = models.BigIntegerField(help_text='Last updated timestamp')
```
why a timestamp instead of a datetimefield?

rjw57 replied:
Because that's what JWP uses and we want to compare apples and apples when determining what's updated.

rjw57 commented:
@abrahammartin OK, I've updated the PR now after your comments. Most notably, I've updated the primary key of the various ``mediaplatform`` models to be an opaque token which is generated via the builtin ``secrets`` module. This should simplify URL encoding since the primary key can be interpolated directly into the URL.

I've also changed the ``on_delete`` behaviour for SMS and JWP models to "SET_NULL". This essentially makes the database neutral on whether there ever can be a video on JWP or the SMS which we shouldn't know about.

The full diff from the original PR is visible here:

https://github.com/rjw57/sms-webapp/compare/issue-121-add-models-v1...rjw57:issue-121-add-models?expand=1
